kubernetes/agent.ts

```diff
@@ -0,0 +1,107 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as kubernetes from "@pulumi/kubernetes";
+
+export interface PulumiSelfHostedAgentComponentArgs {
+    namespace: kubernetes.core.v1.Namespace;
+    imageName: pulumi.Input<string>;
+    selfHostedAgentsAccessToken: pulumi.Input<string>;
+}
+
+export class PulumiSelfHostedAgentComponent extends pulumi.ComponentResource {
+    public readonly agentDeployment: kubernetes.apps.v1.Deployment;
+    constructor(name: string, args: PulumiSelfHostedAgentComponentArgs, opts?: pulumi.ComponentResourceOptions) {
+        super("pulumi-service:kubernetes:PulumiSelfHostedAgentComponentArgs", name, args, opts);
+
+        const agentSecret = new kubernetes.core.v1.Secret("agent-secret", {
+            metadata: {
+                name: "agent-secret",
+                namespace: args.namespace.metadata.name,
+              },
+              stringData: {
+                "PULUMI_AGENT_TOKEN": args.selfHostedAgentsAccessToken,
+              }
+        });
+
+        this.agentDeployment = new kubernetes.apps.v1.Deployment("deployment-agent-pool", {
+            metadata: {
+                name: "deployment-agent-pool",
+                namespace: args.namespace.metadata.name,
+                annotations: {
+                    "app.kubernetes.io/name": "pulumi-deployment-agent-pool",
+                },
+            },
+            spec: {
+                replicas: 1,
+                selector: {
+                    matchLabels: {
+                        app: "pulumi-deployment-agent-pool",
+                    },
+                },
+                template: {
+                    metadata: {
+                        labels: {
+                            app: "pulumi-deployment-agent-pool",
+                            "app.kubernetes.io/name": "pulumi-deployment-agent-pool",
+                        },
+                    },
+                    spec: {
+                        containers: [
+                        {
+                            name: "agent",
+                            image: args.imageName,
+                            imagePullPolicy: "IfNotPresent",
```

Comment:
Do we want if not present over always? If you're using a tag that doesn't change (e.g. `latest`) and the image with this tag is already on the node you'll never pull a newer version (even if one has been published)

Reply:
I will make configurable, as it depends on the imageName and what people want from their setup